File: JAVA/src/main/scala/engine/util/ResourceLoader.scala
package engine.util

import java.io.{InputStreamReader, BufferedReader, InputStream}

import engine.math.Vector3f
import engine.mesh.{Vertex, Mesh}

import scala.collection.mutable.ListBuffer
import scala.io.{BufferedSource, Source}

/**
  * Created by Mnenmenth Alkaborin
  * Please refer to LICENSE file if included
  * for licensing information
  * https://github.com/Mnenmenth
  */
object ResourceLoader {

  def loadShader(fileName: String, isStream: Boolean, path: String = "/engine/shaders/"): String = {
    if(isStream) {
      Source.fromInputStream(getClass.getResourceAsStream(path + fileName)).getLines().mkString("\n")
    }
    else
      Source.fromFile(path + fileName).getLines().mkString("\n")
  }

  def loadMesh(fileName: String, isStream: Boolean, path: String = "/engine/models/"): Mesh = {
    val vertices = new ListBuffer[Vertex]
    val indicies = new ListBuffer[Int]
    def removeEmptyStrings(data: Array[String]): Array[String] = {
      val res = new ListBuffer[String]
      data.foreach(d => if(!d.equals("")) res += d)
      res.toArray
    }

    def read(buffSource: Iterator[String]): Unit ={

      buffSource.foreach{ l =>
        val tokens = l.split(" ")

        if(tokens(0).equals("v")) vertices += new Vertex(new Vector3f(tokens(1).toFloat,
                                                                      tokens(2).toFloat,
                                                                      tokens(3).toFloat))
        else if(tokens(0).equals("f")){
          indicies += tokens(1).toInt - 1
          indicies += tokens(2).toInt - 1
          indicies += tokens(3).toInt -  1
        }
      }

    }

    if(isStream)
      read(Source.fromInputStream(getClass.getResourceAsStream(path + fileName)).getLines())
    else
      read(Source.fromFile(path + fileName).getLines())

    val res = new Mesh
    res.addVertices(vertices.toArray, indicies.toArray)
    res
  }

}

File: CPP/src/engine/util/ResourceLoader.cpp
//
// Created by Razim on 3/2/2016.
//

#include "ResourceLoader.h"
#include "Util.h"
#include <fstream>
#include <iostream>

const char *ResourceLoader::loadShader(std::string fileName, std::string path) {
    std::string shader = "";
    std::ifstream file(path + fileName);

    std::string line;
    if (file.is_open()) {
        while (std::getline(file, line)) {
            shader += (line + "\n");
        }
    } else {
        std::cout << "Shader failed to load" << std::endl;
    }
    return shader.c_str();
}

ResourceLoader::MeshVerts ResourceLoader::loadMesh(std::string fileName, std::string path) {
    std::vector<Vertex> vertices;
    std::vector<int> indices;

    std::string line;
    std::ifstream file(path + fileName, std::ios::in);
    if (file.is_open()) {
        while (std::getline(file, line)) {
            std::vector<std::string> tokens = Util::splitString(line, ' ');
            if (tokens[0] == "v") {
                vertices.push_back(Vertex(Vector3f(std::stof(tokens[1]), std::stof(tokens[2]), std::stof(tokens[3]))));
            } else if (tokens[0] == "f") {
                indices.push_back(std::stoi(tokens[1]) - 1);
                indices.push_back(std::stoi(tokens[2]) - 1);
                indices.push_back(std::stoi(tokens[3]) - 1);
            }
        }
    } else {
        std::cout << "Mesh failed to load" << std::endl;
    }

    MeshVerts mesh;
    mesh.vertices = vertices.data();
    mesh.indices = indices.data();

    return mesh;
}
